jax/experimental/export/_shape_poly.py

```diff
@@ -933,6 +985,119 @@ def cmp_sequence(s1, s2, elem_cmp) -> int:
+
+    constraints = self._parse_constraints(constraints_str)
+    for c, c_str in zip(constraints, constraints_str):
+      if (const := c.to_constant()) is not None:
+        if const < 0:
+          raise ValueError(f"Unsatisfiable explicit constraint: {c_str}")
+        continue
+      self._explicit_constraints.append((c, c_str))
+      self._process_constraint(c, c_str)
+
+  def __str__(self) -> str:
+    extras = []
+    if self._explicit_constraints:
+      extras.append(" with constraints:")
+      for _, c_str in self._explicit_constraints:
+        extras.append(f"  {c_str}")
+    loc = source_info_util._summarize_frame(self._location_frame) if self._location_frame else "unknown"
+    return (
+        f"{id(self)} created at {loc}" +
+        "\n".join(extras))
+  __repr__ = __str__
+
+  def _parse_constraints(self, constraints_str: Sequence[str]) -> Sequence[_DimExpr]:
+    # Parse some contraints into the current scope.
+    def parse_one(cs: str) -> _DimExpr:
+      if not isinstance(cs, str):
+        raise ValueError(
+            f"symbolic_scope must be invoked with a string: got {repr(cs)}")
+      eq_pos = cs.find("=")
+      if eq_pos <= 0 or cs[eq_pos - 1] not in [">", "<"]:
+        raise ValueError("Constraint parsing error: must contain one of '>=' or '<='")
+      e1_str = cs[:eq_pos - 1]
+      e1, = _Parser(e1_str, None, repr(e1_str), self).parse()
+      e2_str = cs[eq_pos + 1:]
+      e2, = _Parser(e2_str, None, repr(e2_str), self).parse()
+      diff = e1 - e2 if cs[eq_pos - 1] == ">" else e2 - e1
+      return _ensure_poly(diff, "symbolic_scope", self)
+
+    if isinstance(constraints_str, str):
+      raise ValueError(
+          "The symbolic constraints should be a sequence of strings. "
+          f"Got {repr(constraints_str)}")
+    return tuple(parse_one(cs) for cs in constraints_str)
+
+  def _process_constraint(self, e: _DimExpr, e_str: str):
+    # Look for the special case m*mon + n >= 0.
+    # Then assert mon >= ceil(n / m) or mon <= floor(n / m)
+    n = m = 0
+    mon = None
+    nr_non_trivial_monomials = 0
+    for _mon, count in e.monomials():
+      if _mon.degree == 0:
+        n = count
+        continue
+      nr_non_trivial_monomials += 1
+      mon = _mon
+      m = count
+
+    if nr_non_trivial_monomials == 1:
+      assert isinstance(mon, _DimMon)
+      bounds = mon.bounds(self)  # This considers default internal constraints, and
+                             # previous external constraints
+      if m > 0:  # mon >= ceil(-n / m)
+        ge = int(np.ceil(- n / m))
+        new_bounds = (max(ge, bounds[0]), bounds[1])
+      else:  # mon <= floor(-n / m)
+        le = int(np.floor(-n / m))
+        new_bounds = (bounds[0], min(le, bounds[1]))
+      if new_bounds[0] > new_bounds[1]:
+        raise ValueError(f"Unsatisfiable constraints: {e_str}")
+      self._monomial_bounds[mon] = new_bounds
+      return
+
+    # The general case, we just remember this geo0
```

Comment:
Is this comment unfinished or does "geo0"  mean >= 0?

Reply:
Fixed